src/Microsoft.TemplateEngine.Cli/Commands/InstantiateCommand.Help.cs

```diff
@@ -0,0 +1,473 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+#nullable enable
+
+using System.CommandLine;
+using System.CommandLine.Help;
+using System.CommandLine.Parsing;
+using System.Diagnostics.CodeAnalysis;
+using System.Text;
+using Microsoft.TemplateEngine.Abstractions;
+using Microsoft.TemplateEngine.Cli.Extensions;
+using Microsoft.TemplateEngine.Cli.TabularOutput;
+using Microsoft.TemplateEngine.Edge.Settings;
+using Microsoft.TemplateEngine.Utils;
+
+namespace Microsoft.TemplateEngine.Cli.Commands
+{
+    internal partial class InstantiateCommand
+    {
+        public void WriteHelp(HelpContext context)
+        {
+            WriteHelp(context, context.ParseResult);
+        }
+
+        public void WriteHelp(HelpContext context, ParseResult parseResult)
+        {
+            InstantiateCommandArgs instantiateCommandArgs = new InstantiateCommandArgs(this, parseResult);
+            if (string.IsNullOrWhiteSpace(instantiateCommandArgs.ShortName))
+            {
+                WriteCustomInstantiateHelp(context);
+                return;
+            }
+
+            IEngineEnvironmentSettings environmentSettings = GetEnvironmentSettingsFromArgs(instantiateCommandArgs);
+
+            using TemplatePackageManager templatePackageManager = new TemplatePackageManager(environmentSettings);
+            HostSpecificDataLoader hostSpecificDataLoader = new HostSpecificDataLoader(environmentSettings);
+
+            //TODO: consider use cache only for help
+            var selectedTemplateGroups = Task.Run(
+                async () => await GetMatchingTemplateGroupsAsync(
+                    instantiateCommandArgs,
+                    templatePackageManager,
+                    hostSpecificDataLoader,
+                    default).ConfigureAwait(false))
+                .GetAwaiter()
+                .GetResult();
+
+            if (!selectedTemplateGroups.Any())
+            {
+                HandleNoMatchingTemplateGroup(instantiateCommandArgs);
+                return;
+            }
+            if (selectedTemplateGroups.Count() > 1)
+            {
+                HandleAmbiguousTemplateGroup(instantiateCommandArgs);
+                return;
+            }
+
+            TemplateGroup templateGroup = selectedTemplateGroups.Single();
+            IEnumerable<TemplateCommand> matchingTemplates =
+                GetMatchingTemplates(
+                    this,
+                    instantiateCommandArgs,
+                    environmentSettings,
+                    templatePackageManager,
+                    templateGroup);
+
+            if (!VerifyMatchingTemplates(
+                environmentSettings,
+                context.Output,
+                matchingTemplates,
+                out IEnumerable<TemplateCommand>? templatesToShow))
+            {
+                //error
```

Comment:
Under what circumstances would help fail?

Reply:
This is related to our "reparsing" thing. 
Examples may be:
`dotnet new bla -h` - `bla` template doesn't exist, we cannot show the help
`dotnet new my-multi-language-template` - we can only show help for single language, and user didn't specify one. 

There are more complicated cases. 

In fact this is minor imo. Previously in these cases `dotnet new` returned error exit code, but now it will not, but still show this message. Also we cannot use colors in help output which is a pity, but also minor.